File: daedalos-tools/journal/lib/collector.py
"""Journal event collector - aggregates events from all Daedalos tools."""

import os
import json
import glob
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from pathlib import Path
from typing import Optional, List, Dict, Any, Iterator


# Data directories
DATA_DIR = Path(os.environ.get("XDG_DATA_HOME", Path.home() / ".local" / "share")) / "daedalos"
JOURNAL_DIR = DATA_DIR / "journal"


@dataclass
class Event:
    """A single event in the journal."""
    timestamp: float
    source: str  # Which tool/component generated the event
    event_type: str  # Type of event (e.g., "loop_started", "file_changed", "gate_check")
    summary: str  # Human-readable summary
    details: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "timestamp": self.timestamp,
            "source": self.source,
            "event_type": self.event_type,
            "summary": self.summary,
            "details": self.details,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "Event":
        return cls(
            timestamp=data.get("timestamp", 0),
            source=data.get("source", "unknown"),
            event_type=data.get("event_type", "unknown"),
            summary=data.get("summary", ""),
            details=data.get("details", {}),
        )


def ensure_journal_dir():
    """Ensure journal directory exists."""
    JOURNAL_DIR.mkdir(parents=True, exist_ok=True)


def collect_gate_events(since: float) -> Iterator[Event]:
    """Collect gate check events."""
    gates_dir = DATA_DIR / "gates"
    if not gates_dir.exists():
        return

    for log_file in sorted(gates_dir.glob("gates-*.jsonl")):
        try:
            with open(log_file) as f:
                for line in f:
                    try:
                        data = json.loads(line)
                        ts = data.get("timestamp", 0)
                        if ts < since:
                            continue

                        result = data.get("result", {})
                        allowed = "allowed" if result.get("allowed") else "denied"
                        gate = data.get("gate", "unknown")
                        source = data.get("source", "unknown")

                        yield Event(
                            timestamp=ts,
                            source="gates",
                            event_type="gate_check",
                            summary=f"Gate '{gate}' {allowed} (from {source})",
                            details=data,
                        )
                    except json.JSONDecodeError:
                        pass
        except Exception:
            pass


def collect_loop_events(since: float) -> Iterator[Event]:
    """Collect loop iteration events."""
    loop_dir = DATA_DIR / "loop"
    if not loop_dir.exists():
        return

    for log_file in sorted(loop_dir.glob("*.json")):
        try:
            with open(log_file) as f:
                data = json.load(f)

            ts = data.get("started", 0)
            if ts < since:
                continue

            task = data.get("task", "unknown task")
            status = data.get("status", "unknown")
            iterations = data.get("iteration", 0)

            yield Event(
                timestamp=ts,
                source="loop",
                event_type="loop_started",
                summary=f"Loop started: {task}",
                details=data,
            )

            # Add end event if completed
            if status in ("completed", "failed", "stopped"):
                ended = data.get("ended", ts)
                yield Event(
                    timestamp=ended,
                    source="loop",
                    event_type=f"loop_{status}",
                    summary=f"Loop {status} after {iterations} iterations: {task}",
                    details=data,
                )
        except Exception:
            pass


def collect_agent_events(since: float) -> Iterator[Event]:
    """Collect agent lifecycle events."""
    agent_dir = DATA_DIR / "agent"
    if not agent_dir.exists():
        return

    # Check for agent activity logs
    for log_file in sorted(agent_dir.glob("*.jsonl")):
        try:
            with open(log_file) as f:
                for line in f:
                    try:
                        data = json.loads(line)
                        ts = data.get("timestamp", 0)
                        if ts < since:
                            continue

                        event_type = data.get("event", "agent_event")
                        agent_name = data.get("name", "unknown")
                        summary = data.get("summary", f"Agent event: {agent_name}")

                        yield Event(
                            timestamp=ts,
                            source="agent",
                            event_type=event_type,
                            summary=summary,
                            details=data,
                        )
                    except json.JSONDecodeError:
                        pass
        except Exception:
            pass


def collect_undo_events(since: float) -> Iterator[Event]:
    """Collect file change events from undo timeline."""
    undo_dir = DATA_DIR / "undo"
    if not undo_dir.exists():
        return

    timeline_file = undo_dir / "timeline.jsonl"
    if not timeline_file.exists():
        return

    try:
        with open(timeline_file) as f:
            for line in f:
                try:
                    data = json.loads(line)
                    ts = data.get("timestamp", 0)
                    if ts < since:
                        continue

                    file_path = data.get("path", "unknown")
                    action = data.get("action", "modified")
                    checkpoint = data.get("checkpoint")

                    if checkpoint:
                        yield Event(
                            timestamp=ts,
                            source="undo",
                            event_type="checkpoint",
                            summary=f"Checkpoint created: {checkpoint}",
                            details=data,
                        )
                    else:
                        yield Event(
                            timestamp=ts,
                            source="undo",
                            event_type=f"file_{action}",
                            summary=f"File {action}: {file_path}",
                            details=data,
                        )
                except json.JSONDecodeError:
                    pass
    except Exception:
        pass


def collect_mcp_events(since: float) -> Iterator[Event]:
    """Collect MCP hub events."""
    mcp_dir = DATA_DIR / "mcp-hub"
    if not mcp_dir.exists():
        return

    for log_file in sorted(mcp_dir.glob("*.log")):
        try:
            stat = log_file.stat()
            if stat.st_mtime < since:
                continue

            # Basic parsing of log files
            with open(log_file) as f:
                for line in f:
                    if "started" in line.lower() or "stopped" in line.lower():
                        yield Event(
                            timestamp=stat.st_mtime,
                            source="mcp-hub",
                            event_type="mcp_event",
                            summary=line.strip()[:100],
                            details={"file": str(log_file)},
                        )
        except Exception:
            pass


def collect_journal_events(since: float) -> Iterator[Event]:
    """Collect events logged directly to the journal."""
    if not JOURNAL_DIR.exists():
        return

    for log_file in sorted(JOURNAL_DIR.glob("journal-*.jsonl")):
        try:
            with open(log_file) as f:
                for line in f:
                    try:
                        data = json.loads(line)
                        ts = data.get("timestamp", 0)
                        if ts < since:
                            continue

                        yield Event.from_dict(data)
                    except json.JSONDecodeError:
                        pass
        except Exception:
            pass


def collect_all_events(
    since: Optional[float] = None,
    hours: float = 24,
    sources: Optional[List[str]] = None,
    event_types: Optional[List[str]] = None,
    limit: int = 1000,
) -> List[Event]:
    """
    Collect events from all sources.

    Args:
        since: Unix timestamp to start from (overrides hours)
        hours: Number of hours to look back (default: 24)
        sources: Filter by source names (e.g., ["gates", "loop"])
        event_types: Filter by event types (e.g., ["gate_check", "file_modified"])
        limit: Maximum number of events

    Returns:
        List of events sorted by timestamp (newest first)
    """
    if since is None:
        since = (datetime.now() - timedelta(hours=hours)).timestamp()

    events = []

    # Collect from all sources
    collectors = {
        "gates": collect_gate_events,
        "loop": collect_loop_events,
        "agent": collect_agent_events,
        "undo": collect_undo_events,
        "mcp-hub": collect_mcp_events,
        "journal": collect_journal_events,
    }

    for source_name, collector in collectors.items():
        if sources and source_name not in sources:
            continue

        for event in collector(since):
            if event_types and event.event_type not in event_types:
                continue
            events.append(event)

    # Sort by timestamp descending (newest first)
    events.sort(key=lambda e: e.timestamp, reverse=True)

    return events[:limit]


def log_event(source: str, event_type: str, summary: str, details: Optional[Dict] = None) -> None:
    """
    Log an event to the journal.

    This can be called by other tools to record events.
    """
    ensure_journal_dir()

    event = Event(
        timestamp=datetime.now().timestamp(),
        source=source,
        event_type=event_type,
        summary=summary,
        details=details or {},
    )

    # Write to daily log file
    date_str = datetime.now().strftime("%Y-%m-%d")
    log_file = JOURNAL_DIR / f"journal-{date_str}.jsonl"

    with open(log_file, "a") as f:
        f.write(json.dumps(event.to_dict()) + "\n")

File: crates/journal/src/collector.rs
//! Event collector - aggregates events from all Daedalos tools
//!
//! Pulls events from:
//! - gates: Gate checks and approvals (JSONL files)
//! - loop: Iteration loop events (JSON state files)
//! - agent: Agent lifecycle events (JSONL logs)
//! - undo: File changes and checkpoints (JSONL timeline)
//! - mcp-hub: MCP server events (log files)
//! - journal: Custom user-logged events (JSONL files)

use crate::db::{Event, JournalDatabase};
use anyhow::Result;
use daedalos_core::Paths;
use serde_json::Value;
use std::collections::HashMap;
use std::fs::{self, File};
use std::io::{BufRead, BufReader};
use std::path::PathBuf;

/// Collector for aggregating events from all sources
pub struct EventCollector {
    data_dir: PathBuf,
}

impl EventCollector {
    /// Create a new event collector
    pub fn new() -> Self {
        let paths = Paths::new();
        Self {
            data_dir: paths.data,
        }
    }

    /// Collect all events from all sources since a given timestamp
    pub fn collect_all(
        &self,
        since: f64,
        sources: Option<&[&str]>,
        event_types: Option<&[&str]>,
        limit: usize,
    ) -> Vec<Event> {
        let mut events = Vec::new();

        // Collect from each source
        let all_sources = ["gates", "loop", "agent", "undo", "mcp-hub", "journal"];

        for source in &all_sources {
            if let Some(filter) = sources {
                if !filter.contains(source) {
                    continue;
                }
            }

            let source_events = match *source {
                "gates" => self.collect_gate_events(since),
                "loop" => self.collect_loop_events(since),
                "agent" => self.collect_agent_events(since),
                "undo" => self.collect_undo_events(since),
                "mcp-hub" => self.collect_mcp_events(since),
                "journal" => self.collect_journal_events(since),
                _ => Vec::new(),
            };

            for event in source_events {
                if let Some(types) = event_types {
                    if !types.contains(&event.event_type.as_str()) {
                        continue;
                    }
                }
                events.push(event);
            }
        }

        // Sort by timestamp descending (newest first)
        events.sort_by(|a, b| b.timestamp.partial_cmp(&a.timestamp).unwrap());

        // Apply limit
        events.truncate(limit);

        events
    }

    /// Collect gate check events from JSONL files
    fn collect_gate_events(&self, since: f64) -> Vec<Event> {
        let mut events = Vec::new();
        let gates_dir = self.data_dir.join("gates");

        if !gates_dir.exists() {
            return events;
        }

        // Read gates-*.jsonl files
        if let Ok(entries) = fs::read_dir(&gates_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().map_or(false, |e| e == "jsonl") {
                    if let Ok(file) = File::open(&path) {
                        let reader = BufReader::new(file);
                        for line in reader.lines().flatten() {
                            if let Ok(data) = serde_json::from_str::<Value>(&line) {
                                let ts = data.get("timestamp")
                                    .and_then(|v| v.as_f64())
                                    .unwrap_or(0.0);

                                if ts < since {
                                    continue;
                                }

                                let allowed = data.get("result")
                                    .and_then(|r| r.get("allowed"))
                                    .and_then(|a| a.as_bool())
                                    .unwrap_or(false);

                                let gate = data.get("gate")
                                    .and_then(|g| g.as_str())
                                    .unwrap_or("unknown");

                                let source = data.get("source")
                                    .and_then(|s| s.as_str())
                                    .unwrap_or("unknown");

                                let status = if allowed { "allowed" } else { "denied" };
                                let summary = format!("Gate '{}' {} (from {})", gate, status, source);

                                let mut event = Event {
                                    timestamp: ts,
                                    source: "gates".to_string(),
                                    event_type: "gate_check".to_string(),
                                    summary,
                                    details: HashMap::new(),
                                };

                                // Add relevant details
                                if let Some(obj) = data.as_object() {
                                    for (k, v) in obj {
                                        event.details.insert(k.clone(), v.clone());
                                    }
                                }

                                events.push(event);
                            }
                        }
                    }
                }
            }
        }

        events
    }

    /// Collect loop iteration events from JSON state files
    fn collect_loop_events(&self, since: f64) -> Vec<Event> {
        let mut events = Vec::new();
        let loop_dir = self.data_dir.join("loop");

        if !loop_dir.exists() {
            return events;
        }

        // Read *.json state files
        if let Ok(entries) = fs::read_dir(&loop_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().map_or(false, |e| e == "json") {
                    if let Ok(contents) = fs::read_to_string(&path) {
                        if let Ok(data) = serde_json::from_str::<Value>(&contents) {
                            let started = data.get("started")
                                .and_then(|v| v.as_f64())
                                .unwrap_or(0.0);

                            if started < since {
                                continue;
                            }

                            let task = data.get("task")
                                .and_then(|t| t.as_str())
                                .unwrap_or("unknown task");

                            let status = data.get("status")
                                .and_then(|s| s.as_str())
                                .unwrap_or("unknown");

                            let iterations = data.get("iteration")
                                .and_then(|i| i.as_u64())
                                .unwrap_or(0);

                            // Start event
                            events.push(Event {
                                timestamp: started,
                                source: "loop".to_string(),
                                event_type: "loop_started".to_string(),
                                summary: format!("Loop started: {}", task),
                                details: json_to_details(&data),
                            });

                            // End event if completed/failed/stopped
                            if matches!(status, "completed" | "failed" | "stopped") {
                                let ended = data.get("ended")
                                    .and_then(|v| v.as_f64())
                                    .unwrap_or(started);

                                events.push(Event {
                                    timestamp: ended,
                                    source: "loop".to_string(),
                                    event_type: format!("loop_{}", status),
                                    summary: format!("Loop {} after {} iterations: {}", status, iterations, task),
                                    details: json_to_details(&data),
                                });
                            }
                        }
                    }
                }
            }
        }

        events
    }

    /// Collect agent lifecycle events from JSONL logs
    fn collect_agent_events(&self, since: f64) -> Vec<Event> {
        let mut events = Vec::new();
        let agent_dir = self.data_dir.join("agent");

        if !agent_dir.exists() {
            return events;
        }

        // Read *.jsonl log files
        if let Ok(entries) = fs::read_dir(&agent_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().map_or(false, |e| e == "jsonl") {
                    if let Ok(file) = File::open(&path) {
                        let reader = BufReader::new(file);
                        for line in reader.lines().flatten() {
                            if let Ok(data) = serde_json::from_str::<Value>(&line) {
                                let ts = data.get("timestamp")
                                    .and_then(|v| v.as_f64())
                                    .unwrap_or(0.0);

                                if ts < since {
                                    continue;
                                }

                                let event_type = data.get("event")
                                    .and_then(|e| e.as_str())
                                    .unwrap_or("agent_event")
                                    .to_string();

                                let agent_name = data.get("name")
                                    .and_then(|n| n.as_str())
                                    .unwrap_or("unknown");

                                let summary = data.get("summary")
                                    .and_then(|s| s.as_str())
                                    .map(|s| s.to_string())
                                    .unwrap_or_else(|| format!("Agent event: {}", agent_name));

                                events.push(Event {
                                    timestamp: ts,
                                    source: "agent".to_string(),
                                    event_type,
                                    summary,
                                    details: json_to_details(&data),
                                });
                            }
                        }
                    }
                }
            }
        }

        events
    }

    /// Collect file change events from undo timeline
    fn collect_undo_events(&self, since: f64) -> Vec<Event> {
        let mut events = Vec::new();
        let undo_dir = self.data_dir.join("undo");
        let timeline_file = undo_dir.join("timeline.jsonl");

        if !timeline_file.exists() {
            return events;
        }

        if let Ok(file) = File::open(&timeline_file) {
            let reader = BufReader::new(file);
            for line in reader.lines().flatten() {
                if let Ok(data) = serde_json::from_str::<Value>(&line) {
                    let ts = data.get("timestamp")
                        .and_then(|v| v.as_f64())
                        .unwrap_or(0.0);

                    if ts < since {
                        continue;
                    }

                    let file_path = data.get("path")
                        .and_then(|p| p.as_str())
                        .unwrap_or("unknown");

                    let action = data.get("action")
                        .and_then(|a| a.as_str())
                        .unwrap_or("modified");

                    let checkpoint = data.get("checkpoint")
                        .and_then(|c| c.as_str());

                    if let Some(cp) = checkpoint {
                        events.push(Event {
                            timestamp: ts,
                            source: "undo".to_string(),
                            event_type: "checkpoint".to_string(),
                            summary: format!("Checkpoint created: {}", cp),
                            details: json_to_details(&data),
                        });
                    } else {
                        events.push(Event {
                            timestamp: ts,
                            source: "undo".to_string(),
                            event_type: format!("file_{}", action),
                            summary: format!("File {}: {}", action, file_path),
                            details: json_to_details(&data),
                        });
                    }
                }
            }
        }

        events
    }

    /// Collect MCP hub events from log files
    fn collect_mcp_events(&self, since: f64) -> Vec<Event> {
        let mut events = Vec::new();
        let mcp_dir = self.data_dir.join("mcp-hub");

        if !mcp_dir.exists() {
            return events;
        }

        // Read *.log files
        if let Ok(entries) = fs::read_dir(&mcp_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().map_or(false, |e| e == "log") {
                    if let Ok(metadata) = fs::metadata(&path) {
                        if let Ok(modified) = metadata.modified() {
                            let mtime = modified
                                .duration_since(std::time::UNIX_EPOCH)
                                .map(|d| d.as_secs_f64())
                                .unwrap_or(0.0);

                            if mtime < since {
                                continue;
                            }

                            // Basic parsing of log files for start/stop events
                            if let Ok(contents) = fs::read_to_string(&path) {
                                for line in contents.lines() {
                                    let line_lower = line.to_lowercase();
                                    if line_lower.contains("started") || line_lower.contains("stopped") {
                                        let summary = line.chars().take(100).collect::<String>();
                                        events.push(Event {
                                            timestamp: mtime,
                                            source: "mcp-hub".to_string(),
                                            event_type: "mcp_event".to_string(),
                                            summary,
                                            details: HashMap::new(),
                                        });
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        events
    }

    /// Collect events logged directly to the journal
    fn collect_journal_events(&self, since: f64) -> Vec<Event> {
        let mut events = Vec::new();
        let journal_dir = self.data_dir.join("journal");

        if !journal_dir.exists() {
            return events;
        }

        // Read journal-*.jsonl files
        if let Ok(entries) = fs::read_dir(&journal_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                let filename = path.file_name()
                    .and_then(|f| f.to_str())
                    .unwrap_or("");

                if filename.starts_with("journal-") && filename.ends_with(".jsonl") {
                    if let Ok(file) = File::open(&path) {
                        let reader = BufReader::new(file);
                        for line in reader.lines().flatten() {
                            if let Ok(data) = serde_json::from_str::<Value>(&line) {
                                let ts = data.get("timestamp")
                                    .and_then(|v| v.as_f64())
                                    .unwrap_or(0.0);

                                if ts < since {
                                    continue;
                                }

                                let source = data.get("source")
                                    .and_then(|s| s.as_str())
                                    .unwrap_or("user")
                                    .to_string();

                                let event_type = data.get("event_type")
                                    .and_then(|t| t.as_str())
                                    .unwrap_or("custom")
                                    .to_string();

                                let summary = data.get("summary")
                                    .and_then(|s| s.as_str())
                                    .unwrap_or("")
                                    .to_string();

                                events.push(Event {
                                    timestamp: ts,
                                    source,
                                    event_type,
                                    summary,
                                    details: json_to_details(&data),
                                });
                            }
                        }
                    }
                }
            }
        }

        events
    }

    /// Sync collected events to the database
    #[allow(dead_code)]
    pub fn sync_to_db(&self, db: &JournalDatabase, since: f64) -> Result<usize> {
        let events = self.collect_all(since, None, None, 10000);
        let mut count = 0;

        for event in events {
            // Note: This is a simple approach - in production, we'd want to
            // deduplicate based on (timestamp, source, event_type, summary)
            db.log_event(&event)?;
            count += 1;
        }

        Ok(count)
    }
}

impl Default for EventCollector {
    fn default() -> Self {
        Self::new()
    }
}

/// Convert a JSON Value to a HashMap for event details
fn json_to_details(value: &Value) -> HashMap<String, Value> {
    let mut details = HashMap::new();
    if let Some(obj) = value.as_object() {
        for (k, v) in obj {
            // Skip the standard fields
            if !matches!(k.as_str(), "timestamp" | "source" | "event_type" | "summary") {
                details.insert(k.clone(), v.clone());
            }
        }
    }
    details
}

/// Log a custom event to the journal JSONL file
pub fn log_event_to_file(source: &str, event_type: &str, summary: &str) -> Result<()> {
    let paths = Paths::new();
    let journal_dir = paths.data.join("journal");
    fs::create_dir_all(&journal_dir)?;

    let event = Event::new(source, event_type, summary);

    // Write to daily log file
    let date_str = chrono::Local::now().format("%Y-%m-%d");
    let log_file = journal_dir.join(format!("journal-{}.jsonl", date_str));

    let mut file = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&log_file)?;

    use std::io::Write;
    writeln!(file, "{}", serde_json::to_string(&event)?)?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_event_creation() {
        let event = Event::new("test", "test_event", "Test summary");
        assert_eq!(event.source, "test");
        assert_eq!(event.event_type, "test_event");
        assert_eq!(event.summary, "Test summary");
        assert!(event.timestamp > 0.0);
    }

    #[test]
    fn test_collector_creation() {
        let collector = EventCollector::new();
        // Should not panic
        let events = collector.collect_all(0.0, None, None, 100);
        // May or may not have events depending on system state
        assert!(events.len() <= 100);
    }
}
